src/Orleans.Core/Timers/TimerManager.cs

```diff
@@ -0,0 +1,363 @@
+using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Orleans.Threading;
+
+namespace Orleans.Timers
+{
+    public static class TimerManager
+    {
+        public static Task Delay(TimeSpan timeout) => DelayUntil(DateTime.UtcNow + timeout);
+
+        public static Task DelayUntil(DateTime dueTime)
+        {
+            var result = new DelayTimer(dueTime);
+            TimerManager<DelayTimer>.Register(result);
+            return result.Completion;
+        }
+
+        private sealed class DelayTimer : ITimerCallback, ILinkedListElement<DelayTimer>
+        {
+            private readonly TaskCompletionSource<int> completion =
+                new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            public DelayTimer(DateTime dueTime)
+            {
+                this.DueTime = dueTime;
+            }
+
+            public Task Completion => this.completion.Task;
+
+            public DateTime DueTime { get; }
+
+            public bool IsCanceled => false;
+
+            public void OnTimeout() => this.completion.TrySetResult(0);
+
+            DelayTimer ILinkedListElement<DelayTimer>.Next { get; set; }
+        }
+    }
+
+    /// <summary>
+    /// Manages timers of a specified type, firing them after they expire.
+    /// </summary>
+    /// <typeparam name="T">The timer type.</typeparam>
+    internal static class TimerManager<T> where T : class, ITimerCallback, ILinkedListElement<T>
+    {
+        /// <summary>
+        /// The maximum number of times a queue can be denied servicing before servicing is mandatory.
+        /// </summary>
+        private const int MaxStarvation = 2;
```

Comment:
`TimerManager<T>` is static at the moment, we'd have to change that before making it configurable

Reply:
That's fine for now then.

Reply:
It would be reasonable to want to configure `TimerTickMilliseconds`. `TimerTickMilliseconds` could also be made dynamic, but I figured that was too much of a change for now.

The NuGet package used to originally test the theory used a dedicated thread instead of a `Timer`.